### ISSUE TITLE
Passing block param in component - modified twice in a single render

### ISSUE THREAD

**Issue by typeoneerror** [emberjs/ember.js#11519]
I'm not sure if this is related to https://github.com/emberjs/ember.js/issues/11277, but I'm experiencing a similar issue. I've been able to reproduce the issue on [this jsbin](http://emberjs.jsbin.com/pidufugama/1/edit?html,js,output).

The first two uses of {{editable-thing}} work fine, but the last where we pass through the `isEditing` param via yield starts the unending warning loop (fair warning, open Chrome's [task manager](http://stackoverflow.com/a/27261978/53653) so you can kill the "runaway process" when you click the third edit button). 

Is this a bug or a misuse/misunderstanding of block params? Either way, would love to understand this better.

**Comment by Samsinite** [comment]
Looks just like what I was seeing in #11277, @stefanpenner @rwjblue any ideas?

**Comment by sandstrom** [comment]
We're seeing something very similar in our app.

**Comment by sergetab** [comment]
Same problem here.

**Comment by stefanpenner** [comment]
> Looks just like what I was seeing in #11277, @stefanpenner @rwjblue any ideas?

no need to ping me, i review every notifcation

**Comment by Samsinite** [comment]
Sorry I wasn't sure. #11277 seemed to sneak into 1.13 even though I reported 28 days ago.

**Comment by typeoneerror** [comment]
@Samsinite I'm positive there are limited hands to review and fix bugs, so my guess is the core team prioritizes fixes. I'm guessing this means that not all issues get fixed in each 6-week release cycle. 

If I could make a suggestion, I did notice that your issue was perhaps harder to verify bc you had to clone a whole repo and run a project locally to test. Maybe that slowed the review process?

**Comment by Samsinite** [comment]
Possibly, thanks for duplicating the issue into a jsbin. I am guessing that the core team is a little swamped right now when looking at the number of open issues on Github.

Didn't mean to sound ungrateful as the ember core team does awesome work, just wanted them to be aware of the issue.

**Comment by typeoneerror** [comment]
@Samsinite no worries. want to close your issue and roll with this one with the jsbin?

**Comment by Samsinite** [comment]
Sounds good, this one is newer so it likely has more visibility too.

**Comment by ammorrise** [comment]
I'm having this same issue.

**Comment by jmurphyau** [comment]
TLDR;

Using the variable being yielded in the component template causes a loop..

Using `{{unbound [var]}}` and manually call `this.rerender()` works around the issue for the time being.

------------

I looked into it and this is what I could find - might help someone with more experience solve this problem.. 

First Render:

1. `yield` constructs the block
2. `RenderResult.build`
3. `RenderResult.bindLocals(blockArguments)`
4. `[ember-htmlbars] env.bindLocal` (for each blockArgument)
5. `stream.notify()` (for each blockArgument)

Subsequent Renders (value changes):

1. `yield` constructs the block
2. `RenderResult.revalidateWith`
3. `RenderResult.updateLocals(blockArguments)`
4. `[htmlbars] env.updateLocal` (for each blockArgument)
5. `[ember-htmlbars] env.bindLocal` (for each blockArgument - `updateLocal` just calls `bindLocal`)
6. `stream.notify()` (for each blockArgument)

Step 6 on rerender is where the issue starts. This call to `notify` triggers the component to rerender, which causes it to call `yield` again and start from Step 1. 

Looks like this is because the component is subscribed to this particular stream, and it's subscribed to the stream because it's used in the template body of the component.. 

Changing the variable used in the component tempalte body to `{{unbound [var]}}` doesn't cause this issue but doesn't update the template..

Commenting out [this line](https://github.com/emberjs/ember.js/blob/680f997ed0958c420abdcd0b1673111aee26afe7/packages/ember-htmlbars/lib/hooks/bind-local.js#L19) fixes the issue - but I'm not sure what the side effect would be. 

Someone with more knowledge on HTMLBars would be able to best come up with a solution I think.

[This JSBin](http://jsbin.com/ralivu/edit?html,js,output) is the bare minimum needed to get this issue to occur.
[This JSBin](http://jsbin.com/sariqa/edit?html,js,output) uses `{{unbound}}` and doesn't have the issue but manually calling `this.rerender()` isn't ideal.. Might get people by for the time being as a work around.

**Comment by jmurphyau** [comment]
A better workaround is to create a computed property that is used just for yielding that produces the same value.. E.g:

`block-with-yield.js`

    App.BlockWithYieldComponent = Ember.Component.extend({
      danger: 0,
      dangerYield: Ember.computed.oneWay('danger'),
      actions: {
        updateDangerValue: function() {
          this.set('danger', 2);
        }
      }
    });

`block-with-yield.hbs`

    <p>
      This text line is inside the component
      [danger = {{danger}}] 
      [<a href='#' {{action 'updateDangerValue'}}>Update Danger Value</a>]
      {{yield dangerYield}}
    </p>


[JSBin](http://jsbin.com/qoyero/edit?html,js,output)

**Comment by tomdale** [comment]
Looking into this now.

**Comment by tomdale** [comment]
The issue is that Glimmer is re-using the yielded stream from the shadow scope instead of creating a proxy stream. When it re-validates the light scope template, it triggers a change notification on the block param stream. Unfortunately, this is the same stream from the shadow scope, and because `danger` is yielded in in the example above, it causes an infinite loop of the shadow scope and light scope continuously invalidating one another.

**Comment by tomdale** [comment]
```js
var newValue = Stream.wrap(value, ProxyStream, key);
```

I believe this line is the culprit. Reading this code, you may assume that the value is *always* wrapped in a `ProxyStream`, which is what I believe the original author intended. Unfortunately, that's only true if the value is not a stream (`Stream.wrap` returns the identity if `isStream` is true).

**Comment by tomdale** [comment]
I think the call to `existing.notify()` is unnecessary as well, as @jmurphyau pointed out. In theory we only need to invalidate the stream if the source changes, and indeed, calling `setSource` on a stream will automatically `notify()`.

The question here is whether we need to create a new `ProxyStream` for each block param. In theory, we could re-use the value passed in. If it's a stream, we'll get notified when it changes. If it's a primitive, we can wrap it and update it whenever the new primitive comes in.

I can't come up with a reason why we have to force revalidation of the block param stream on re-renders, but maybe I'm forgetting something. It seems like a performance win if we can avoid it.

**Comment by Samsinite** [comment]
Rad, thanks guys! This will go in a 1.13.x release before 2.0 right?

**Comment by rwjblue** [comment]
@Samsinite - Yes, absolutely.

**Comment by sandstrom** [comment]
I can confirm that this solved our issue! Thanks @jmurphyau and @tomdale :boat:
